File: src/rust/src/command.rs

#[derive(Debug, PartialEq)]
pub enum Command {
    IncPtr,
    DecPtr,
    IncVal,
    DecVal,
    PutChar,
    GetChar,
    OpenBr,
    CloseBr,
}

impl Command {
    pub fn from_char(ch: char) -> Option<Self> {
        match ch {
            '>' => Some(Command::IncPtr),
            '<' => Some(Command::DecPtr),
            '+' => Some(Command::IncVal),
            '-' => Some(Command::DecVal),
            '.' => Some(Command::PutChar),
            ',' => Some(Command::GetChar),
            '[' => Some(Command::OpenBr),
            ']' => Some(Command::CloseBr),
            _ => None
        }
    }
}

pub fn parse(code: &str) -> Vec<Command> {
    return code.chars().filter_map(|ch| Command::from_char(ch)).collect();
}

//pub fn verify(code: &Vec<Command>) -> Result<(), &'static str> {
//    let mut bracket_bal = 0;
//    for com in code {
//        match com {
//            Command::OpenBr => {bracket_bal += 1;},
//            Command::CloseBr => {bracket_bal -= 1;},
//            _ => {}
//        }
//    }
//    if bracket_bal != 0 {
//        return Err("Unbalanced Brackets")
//    }
//    Ok(())
//}

//pub fn execute_prg(prg: &Vec<Command>) {
//    let mut prg_head = 0;
//    let mut mem: Vec<u8> = vec![0];
//    let mut mem_ptr = 0;
//
//    while prg_head < prg.len() {
//        match prg[prg_head] {
//            Command::IncPtr => {mem_ptr += 1; if mem_ptr == mem.len() {mem.push(0)}},
//            Command::DecPtr => mem_ptr -= 1,
//            Command::IncVal => mem[mem_ptr] = mem[mem_ptr].wrapping_add(1),
//            Command::DecVal => mem[mem_ptr] = mem[mem_ptr].wrapping_sub(1),
//            Command::PutChar => print!("{}", char::from_u32(mem[mem_ptr] as u32).unwrap()),
//            Command::GetChar => (),
//            Command::OpenBr => {
//                if mem[mem_ptr] == 0 {
//                    let mut bracket_bal = 1;
//                    while bracket_bal > 0 {
//                        prg_head += 1;
//                        if prg[prg_head] == Command::OpenBr { bracket_bal += 1; }
//                        if prg[prg_head] == Command::CloseBr { bracket_bal -= 1; }
//                    }
//                }
//            },
//            Command::CloseBr => {
//                let mut bracket_bal = 1;
//                while bracket_bal > 0 {
//                    prg_head -= 1;
//                    if prg[prg_head] == Command::OpenBr { bracket_bal -= 1; }
//                    if prg[prg_head] == Command::CloseBr { bracket_bal += 1; }
//                }
//                prg_head -= 1;
//            },
//        }
//        prg_head += 1;
//    }
//}

File: interpreters/csharp/Command.cs

class Command {
  public static List<Command> Parse(string code) {
    List<Command> result = new List<Command>();
    foreach (char ch in code) {
      switch (ch) {
        case '>':
          result.Add(new IncPtr());
          break;
        case '<':
          result.Add(new DecPtr());
          break;
        case '+':
          result.Add(new IncVal());
          break;
        case '-':
          result.Add(new DecVal());
          break;
        case '.':
          result.Add(new PutChar());
          break;
        case ',':
          result.Add(new GetChar());
          break;
        case '[':
          result.Add(new OpenBr());
          break;
        case ']':
          result.Add(new CloseBr());
          break;
      }
    }
    return result;
  }

  public static void Execute(List<Command> program) {
    byte[] memory = new byte[30000];
    int mem_head = 0;
    int prg_head = 0;

    while (prg_head < program.Count) {
      switch (program[prg_head]) {
        case IncPtr _:
          mem_head++;
          break;
        case DecPtr _:
          mem_head--;
          break;
        case IncVal _:
          memory[mem_head]++;
          break;
        case DecVal _:
          memory[mem_head]--;
          break;
        case PutChar _:
          Console.Write((char)memory[mem_head]);
          break;
        case GetChar _:
          // TODO
          break;
        case OpenBr _:
          if (memory[mem_head] == 0) {
            int bracket_bal = 1;
            while (bracket_bal > 0) {
              prg_head++;
              if (program[prg_head] is OpenBr) {
                bracket_bal++;
              } else if (program[prg_head] is CloseBr) {
                bracket_bal--;
              }
            }
          }
          break;
        case CloseBr _:
          if (memory[mem_head] != 0) {
            int bracket_bal = 1;
            while (bracket_bal > 0) {
              prg_head--;
              if (program[prg_head] is CloseBr) {
                bracket_bal++;
              } else if (program[prg_head] is OpenBr) {
                bracket_bal--;
              }
            }
          }
          break;
      }
      prg_head++;
    }
  }
}

class IncPtr: Command {}
class DecPtr: Command {}
class IncVal: Command {}
class DecVal: Command {}
class PutChar: Command {}
class GetChar: Command {}
class OpenBr: Command {}
class CloseBr: Command {}

